Python/Product/PythonTools/PythonTools/Intellisense/AnalysisEntryService.cs

```diff
@@ -0,0 +1,198 @@
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Differencing;
+using Microsoft.VisualStudio.Text.Editor;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.PythonTools.Intellisense {
+    public interface IAnalysisEntryService {
+        bool TryGetAnalyzer(ITextBuffer textBuffer, out ProjectAnalyzer analyzer, out string filename);
+        bool TryGetAnalyzer(ITextView textView, out ProjectAnalyzer analyzer, out string filename);
+
+        Task WaitForAnalyzerAsync(ITextBuffer textBuffer, CancellationToken cancellationToken);
+        Task WaitForAnalyzerAsync(ITextView textView, CancellationToken cancellationToken);
+
+        ProjectAnalyzer DefaultAnalyzer { get; }
+    }
+
+    [Export(typeof(IAnalysisEntryService))]
+    [Export(typeof(AnalysisEntryService))]
+    class AnalysisEntryService : IAnalysisEntryService {
+        private readonly IServiceProvider _site;
+        private readonly IComponentModel _model;
+        private readonly IWpfDifferenceViewerFactoryService _diffService;
+
+        private static readonly object _waitForAnalyzerKey = new object();
+
+        [ImportingConstructor]
+        public AnalysisEntryService([Import(typeof(SVsServiceProvider))] IServiceProvider site) {
+            _site = site;
+            _model = _site.GetComponentModel();
+
+            try {
+                _diffService = _model.GetService<IWpfDifferenceViewerFactoryService>();
+            } catch (CompositionException) {
+            } catch (ImportCardinalityMismatchException) {
+            }
+        }
+
+        public void SetAnalyzer(ITextBuffer textBuffer, VsProjectAnalyzer analyzer) {
+            textBuffer.Properties[typeof(VsProjectAnalyzer)] = analyzer;
+
+            TaskCompletionSource<object> tcs;
+            if (textBuffer.Properties.TryGetProperty(_waitForAnalyzerKey, out tcs)) {
+                tcs.TrySetResult(null);
+                textBuffer.Properties.RemoveProperty(_waitForAnalyzerKey);
+            }
+        }
+
+        /// <summary>
+        /// Gets the analysis entry for the given view and buffer.
+        /// 
+        /// For files on disk this is pretty easy - we analyze each file on it's own in a buffer parser.  
+        /// Therefore we map filename -> analyzer and then get the analysis from teh analyzer.  If we
+        /// determine an analyzer but the file isn't loaded into it for some reason this would return null.
+        /// We can also apply some policy to buffers depending upon the view that they're hosted in.  For
+        /// example if a buffer is outside of any projects, but hosted in a difference view with a buffer
+        /// that is in a project, then we'll look in the view that has the project.
+        /// 
+        /// For interactive windows we will use the analyzer that's configured for the window.
+        /// </summary>
+        public bool TryGetAnalysisEntry(ITextView textView, ITextBuffer textBuffer, out AnalysisEntry entry) {
```

Comment:
>TryGetAnalysisEntry [](start = 20, length = 19)

You use these a lot, but don't expose them in IAnalysisEntryService, and I'm not sure why.